### ISSUE TITLE
borg slow even with no changed files

### ISSUE THREAD

**Issue by bertnp** [borgbackup/borg#4229]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Issue

#### Your borg version (borg -V).
borg 1.1.8

Edit: upgraded from 1.1.5 to 1.1.8 with no improvement

#### Operating system (distribution) and version.
Ubuntu 18.04

#### Hardware / network configuration, and filesystems used.
Somewhat unusual configuration:
Running borg in an Ubuntu VM (using VirtualBox) on a Windows 7 host.
Source filesystem: VirtualBox shared folder
Destination filesystem: ext4 mounted as a raw disk in VirtualBox

#### How much data is handled by borg?
~90k files totalling ~420GB in size

#### Full borg commandline that lead to the problem (leave away excludes and passwords)
`borg create --stats --progress --files-cache=ctime,size --patterns-from <patterns file> <myrepo>::<myarc> source1 source2`

## Describe the problem you're observing.
Backing up takes about 15 minutes, even if I re-run the same backup command (with a different archive name) right after the previous one finishes so no files have changed. Is this a normal time? 

As a point of reference, another (Python) script that I use in my backup workflow also scans the same file tree and performs os.stat() on each file, and completes in about 90 seconds.

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

I have inode checking disabled since VirtualBox shared folders do not appear to have stable inodes.

I have tried listing the output with --list, and all files are marked as unchanged ('U'), so it appears as though it is taking borg 15 minutes just to traverse the file tree.

#### Include any warning/errors/backtraces from the system logs
borg benchmark crud output:
```
C-Z-BIG          23.27 MB/s (10 * 100.00 MB all-zero files: 42.97s)
R-Z-BIG          69.02 MB/s (10 * 100.00 MB all-zero files: 14.49s)
U-Z-BIG          54.79 MB/s (10 * 100.00 MB all-zero files: 18.25s)
D-Z-BIG          98.54 MB/s (10 * 100.00 MB all-zero files: 10.15s)
C-R-BIG          31.95 MB/s (10 * 100.00 MB random files: 31.30s)
R-R-BIG          41.03 MB/s (10 * 100.00 MB random files: 24.37s)
U-R-BIG          75.46 MB/s (10 * 100.00 MB random files: 13.25s)
D-R-BIG          36.62 MB/s (10 * 100.00 MB random files: 27.30s)
C-Z-MEDIUM       16.40 MB/s (1000 * 1.00 MB all-zero files: 60.96s)
R-Z-MEDIUM      106.99 MB/s (1000 * 1.00 MB all-zero files: 9.35s)
U-Z-MEDIUM      156.67 MB/s (1000 * 1.00 MB all-zero files: 6.38s)
D-Z-MEDIUM      557.72 MB/s (1000 * 1.00 MB all-zero files: 1.79s)
C-R-MEDIUM       47.45 MB/s (1000 * 1.00 MB random files: 21.08s)
R-R-MEDIUM       68.16 MB/s (1000 * 1.00 MB random files: 14.67s)
U-R-MEDIUM      234.76 MB/s (1000 * 1.00 MB random files: 4.26s)
D-R-MEDIUM       36.78 MB/s (1000 * 1.00 MB random files: 27.19s)
C-Z-SMALL         2.17 MB/s (10000 * 10.00 kB all-zero files: 46.18s)
R-Z-SMALL        31.16 MB/s (10000 * 10.00 kB all-zero files: 3.21s)
U-Z-SMALL         4.92 MB/s (10000 * 10.00 kB all-zero files: 20.30s)
D-Z-SMALL        53.97 MB/s (10000 * 10.00 kB all-zero files: 1.85s)
C-R-SMALL         1.92 MB/s (10000 * 10.00 kB random files: 51.99s)
R-R-SMALL        32.24 MB/s (10000 * 10.00 kB random files: 3.10s)
U-R-SMALL         4.45 MB/s (10000 * 10.00 kB random files: 22.47s)
D-R-SMALL        28.71 MB/s (10000 * 10.00 kB random files: 3.48s)
```

**Comment by infectormp** [comment]
@bertnp try solution from https://github.com/borgbackup/borg/issues/4192 and 1.1.8 already released

**Comment by bertnp** [comment]
> @bertnp try solution from #4192 and 1.1.8 already released

If understand correctly, the problem in issue #4192 was that the file cache was not history was not long enough which resulted in the need to re-chunk data. In the case I'm reporting, files are listed as unchanged, which would indicate that the problem is not with file cache history length, right?

I'll try again with 1.1.8 (I had installed borg from Ubunutu's repositories which currently have 1.1.5)

**Comment by infectormp** [comment]
You have unstable fs so BORG_FILES_CACHE_TTL can help you in this case. https://borgbackup.readthedocs.io/en/stable/faq.html#it-always-chunks-all-my-files-even-unchanged-ones

**Comment by bertnp** [comment]
> You have unstable fs so BORG_FILES_CACHE_TTL can help you in this case. 

I mount everything in the same place each time, so I think the only unstable part would be the inodes, but I'm using `--file-cache=ctime,size`, so unstable inodes shouldn't cause problems.

Also, just tried backing up again after upgrading from 1.1.5 to 1.1.8 with no improvement.

**Comment by infectormp** [comment]
If you right BORG_FILES_CACHE_TTL will not help you. It remains only to try.

**Comment by bertnp** [comment]
> If you right BORG_FILES_CACHE_TTL will not help you. It remains only to try.

Just tried using `export BORG_FILES_CACHE_TTL=50` with no improvement. I use this repo for only one set of files with the same mount points each time, so 50  should be more than enough to rule out file cache length as a culprit.

**Comment by infectormp** [comment]
Great. Now you can do some more tests. Can you do same test but with local fs, not with VirtualBox shared folder?

**Comment by bertnp** [comment]
Testing on the local fs is tricky -- it's quite small and doesn't have many files to test on.

Alternatively, does borg have any granular performance metrics that can be enabled to determine the specific bits of code that are taking so long to run?

**Comment by ThomasWaldmann** [comment]
See #2352 - not sure you want to run your backups over that.

**Comment by bertnp** [comment]
Thanks for pointing out this out @ThomasWaldmann! The configuration I've described here isn't my only backup, so I'm still curious to see if I can make it work. Aside from the speed issue, it seems to be okay -- I've created several archives and run `borg check` without any problems.

It's also probably worth pointing out that my configuration is different than the one in #2352 -- I am using a (read-only) VirtualBox share as the data source, but backing up to a proper ext4 partition whereas that issue was storing the borg repo in a VirtualBox shared folder.

**Comment by ThomasWaldmann** [comment]
I've seen the difference in the configuration, but your case might be even worse from a safety perspective:
- If borg blows up rather quickly, you know something is wrong and you'll solve it otherwise (like in the borg repo on vboxsf case).
- It would be worse if it just misbehaves without you noticing it, e.g. if the source data read by borg is corrupted silently by vboxsf. And as there is no checking mechanism at the source, you'ld only notice that later.

Not sure if that can actually be, but I'ld be very careful at least and better double check manually.

**Comment by bertnp** [comment]
Indeed, the possibility of the intermediate layer (vboxsf in this case) silently corrupting data is alarming. What is the preferred method for using borg on Windows then? I see that cygwin and WSL are listed as experimental options in the docs, but one could imagine a similar scenario as you've described but exchanging "vboxsf" for "cygwin" or "WSL".

What would you suggest as a manual checking method? Compute file checksums on the source (in Windows, not through vboxsf) and then compare to the backup using `borg mount`?

**Comment by ThomasWaldmann** [comment]
Well, it's a difference whether one can just speculate about potential issues or has really seen some:
- vboxsf: seen issue if repo is on it, we have a ticket about it
- cygwin: iirc, there were issues seen on 32bit, on 64bit it worked, we have a ticket about it
- wsl: microsoft did not implement some sync api call we use in borg 1.1, we have a ticket about it

Yes, computing some sha512sums (of b2sums) directly on windows and just adding them to the backup would help determining the cause in case there is some issue. Guess I'ld just do some major backup and check the stuff right after extracting it again.

**Comment by bertnp** [comment]
I will verify the backup manually and report back. Then this ticket can (hopefully) serve as one example of borg working correctly with vboxsf, aside from the speed issue which this ticket was originally reporting.

Getting back to the speed issue, can someone verify that 15 minutes is abnormally slow for a backup of around 90k files that have not changed (and are registered as unchanged and therefore not being re-chunked)?

If this is indeed slower than normal, what would be a good way to diagnose the cause of the slow speeds? For example, perhaps some strategically placed time.perf_counter() and print statements would help? I can try doing this, but as I'm not familiar with the code base, it would be helpful if someone could point me to the right place to start.

**Comment by Lky** [comment]
Hello i have a similar issue, after upgrade from 1.0.9 (Debian Jessie) to 1.1.8 (binary) in 2019.

I should upgrade earlier, so i could talk with you at 35c3 about this.

have 2 backup jobs
1.0.9 = 2 mins / 1.1.8 = 25 mins
1.0.9 = 12 mins / 1.1.8 = 150 mins

normally i do every hour backups, but because of the long backup time now i can't.
If i get back to work next week, i could give you some logs about it and the command line i use.

**Comment by infectormp** [comment]
@Lky you have this issue with VirtualBox?

**Comment by ThomasWaldmann** [comment]
@Lky as documented in the changelog, the first run is expected to be slow.

do you have slowness for subsequent runs on same source data files?

**Comment by Lky** [comment]
source (ext3) -> ssh -> backup location (ext4)
borg create -C lz4 --exclude-caches --exclude-from "${BACKUP_EXCLUDE}" -v -s borg@${SERVER}:data::'{hostname}_{utcnow:%Y-%m-%d_%H%M%S}' /source (3 dirs)

First run after switch from 1.0.9 to 1.1.8.
Duration: 4 hours 18 minutes 11.14 seconds / Files: 554747 / 175.21 GB
Chunks: 1039359 unique/56809737 total

1.0.9 = Duration: 11 minutes 15.39 seconds / 554742 files
1.1.8 = Duration: 2 hours 29 minutes 17.52 seconds / 554358 files

I had two source jobs, /data (the log above) and /data/group (a subset of the data) and push to the same repo. Subset borg job goes from 2 mins to 25 mins.

So no VirtualBox or anything else, i just switched the versions, allow the repo to push and do the first run. I did read the upgrade notes, but maybe i miss there something.

**Comment by ThomasWaldmann** [comment]
@lky can you copy and paste your posts to a new issue? it might be a different problem than the one this issue was opened for, so let's better not mix them together.

**Comment by ThomasWaldmann** [comment]
@bertnp any news?

**Comment by ThomasWaldmann** [comment]
speed: try `--nobsdflags`.

**Comment by bertnp** [comment]
After making two changes, I've noticed an improvement in speed to 5min 40sec down from ~15min:
- moved the destination backup drive to another machine which is running Ubuntu natively (instead of inside of a VM). This machine is located off-site and accessed remotely via the internet. Backups are still done from within a VM accessing the source files via a VirtualBox shared directory. 
- I've recently switched antivirus software on the Windows host

Unfortunately both of these changes were made without a test backup in between, so I'm not sure which one (if either) are responsible for the speed improvement.

As suggested, I also tried `--nobsdflags` which further sped up backup time to ~4min 20sec.

Still not particularly fast, but a big improvement.

**Comment by Arcovion** [comment]
> Getting back to the speed issue, can someone verify that 15 minutes is abnormally slow for a backup of around 90k files that have not changed (and are registered as unchanged and therefore not being re-chunked)?

I'm also running borg via VirtualBox (`borg 1.1.7` on `ubuntu 18.10`), here's ~155k files totalling 93GB:
```
Duration: 1 hours 12 minutes 26.89 seconds
Number of files: 154658
...
                       Original size      Compressed size    Deduplicated size
This archive:               93.03 GB             76.70 GB            458.15 kB
...
776.84s user 431.29s system 27% cpu 1:12:36.98 total
```
It takes that long even if I re-run the exact same `create` command immediately after. I also tried the suggested `files-cache` argument to avoid checking `inodes`. 
Running `find -type f -printf '%p\t%t\t%s\t%i\n'` in the same location takes:
```
1.88s user 31.35s system 32% cpu 1:43.79 total
```
To find all the filenames including `ctime`/`size`/`inodes` as TSV.
I know that using `find` is much faster than iterating using functions/methods in python (or whatever borg does), but clearly that's not the main issue.

If there *is* a cache, I would expect the filenames to be eagerly loaded and removed from the backup process right from the start, but it's clear with `--progress` that borg is still iterating every single file.
Therefore I'm assuming the files cache is currently somehow incompatible with `vboxsf`. How can I verify if it's working as intended? I haven't done any digging to figure this out yet, so I have no idea where the files cache lives/how it's supposed to operate etc.

Using `--exclude` works just fine, so in theory I could work around it with something like `find` [∆](https://en.wikipedia.org/wiki/Symmetric_difference) `files-cache` > `.borg_blacklist`... but obviously I'd be managing the cache myself at that point and only making partial archives at that.

**Comment by ThomasWaldmann** [comment]
I'ld guess find not being significantly faster for what it does compare with the same sub-functionality of borg. But borg has a bit more to do than just discover the filenames.

Borg always does full backups that logically include ALL the files, so it has to look at all files, whether they have changed or not:
- If a file did not change, borg will create the entry in the backup archive for that file rather quickly, from cached chunk id list but it will still read some metadata from the filesystem (xattrs, acls, bsdflags).
- If a file changed, it will read the file contents, chunk, hash, dedup them. If a new chunk is found, it will compress, encrypt, authenticate, store it.

There's nothing special borg is doing, so it should work with any correctly implemented filesystem.
Not all filesystems are implemented correctly though, some do not even work reliably...
Some might need tuning, e.g. if inodes are instable or ctime/mtime does not behave as expected.

You can find more details about how borg works in the docs (e.g. in "internals" section).

Combining find and borg isn't helpful usually as you would not get the usual full archive.

TLDR: don't use crappy filesystems to do your backups.

**Comment by bertnp** [comment]
@Arcovion The times you are reporting are quite a bit longer than the ones I'm seeing, even after accounting for the fact that you have more files than I do. Have you tried using the `--list` option with `borg create`? All unchanged files should be marked as 'U'. If they are not, this is likely the cause for the long run times as borg will need to rechunk files that are not registered as unchanged.

@ThomasWaldmann As I mentioned earlier, a possibly useful point of reference is another script that I'm running which is able to traverse the same file tree and run `os.stat()` on each file in around 90 seconds, quite a bit faster than `borg create` is taking. You pointed out that borg queries additional metadata beyond what `os.stat()` is doing. Could you tell me (or point me to the right piece of code) which functions borg uses to read metadata from files which have not changed? I can code up a script that traverses the file tree and runs each of the functions that borg does to try to get an idea of what in particular is causing the slow run times. If it is in fact an issue with the "crappy" vboxfs file system, then at least we would have a better idea of why. If not, that would be good to know about too.

**Comment by ThomasWaldmann** [comment]
@bertnp there you go: https://github.com/borgbackup/borg/blob/1.1.9/src/borg/archive.py#L880

(the methods called are right above the linked one)

Besides these fs accesses, borg also makes sure that the chunks that a unchanged file uses are actually still there, so that is a files cache hashtable lookup to find the file's entry and N chunks index hashtable lookups to check whether all the chunks are there (as expected).

**Comment by Arcovion** [comment]
@ThomasWaldmann Thanks. For the record I guess `vboxsf` doesn't support `xattrs`:
```
lsattr: Inappropriate ioctl for device While reading flags on ./
```
I also tried the SMB and NFS file systems, but both were slower and NFS had this error after around 30GB:
```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4434, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 4366, in run
    return set_ec(func(args))
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 152, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 554, in do_create
    create_inner(archive, cache)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 518, in create_inner
    read_special=args.read_special, dry_run=dry_run, st=st)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 631, in _process
    read_special=read_special, dry_run=dry_run)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 631, in _process
    read_special=read_special, dry_run=dry_run)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 631, in _process
    read_special=read_special, dry_run=dry_run)
  [Previous line repeated 2 more times]
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 605, in _process
    status = archive.process_file(path, st, cache)
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 1028, in process_file
    self.chunk_file(item, cache, self.stats, backup_io_iter(self.chunker.chunkify(fd, fh)))
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 960, in chunk_file
    from_chunk, part_number = self.write_part_file(item, from_chunk, part_number)
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 938, in write_part_file
    self.write_checkpoint()
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 447, in write_checkpoint
    self.save(self.checkpoint_name)
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 486, in save
    self.repository.commit()
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 430, in commit
    self.compact_segments()
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 712, in compact_segments
    new_segment, offset = self.io.write_put(key, data, raise_full=True)
  File "/usr/lib/python3/dist-packages/borg/repository.py", line 1459, in write_put
    fd.write(b''.join((crc, header, id, data)))
  File "src/borg/platform/linux.pyx", line 251, in borg.platform.linux.SyncFile.write
OSError: [Errno 5] Input/output error
```
Could be that my mount options were suboptimal, but maybe it's worth hardening borg against these potential errors [with a retry clause like this?](https://stackoverflow.com/a/30630361)
I'm not really invested in that issue though as I won't be using NFS myself.

@bertnp D'oh! So I knew I had to ignore inodes with `--files-cache=mtime,size` because I remembered that had made it faster before... but for some reason that argument was missing when I made the aforementioned archive. My mistake.

New results, with the correct `--files-cache=mtime,size --noatime --noctime --nobsdflags` options:
```
Duration: 10 minutes 48.32 seconds
Number of files: 154658
...
                       Original size      Compressed size    Deduplicated size
This archive:               93.03 GB             76.70 GB             11.60 MB
...
29.82s user 178.70s system 31% cpu 10:56.72 total
```
This shows the files cache does work with `vboxsf` when you ignore `inodes`.
It would be nice to see the total number of file cache hits with `--stats` so you could tell how many files were (not) changed at a glance. I will open an issue for that, as it would have helped me out here.

**Comment by ThomasWaldmann** [comment]
@arcovion i guess @bertnp already was at that conclusion in december (excluding inode from change detection), but still used ctime instead of mtime. Maybe you should consider/try that also, because ctime is safer than mtime (at least for normal linux filesystems).

To test that, you need to run borg at least twice after changing to it.

Your timings seem that this is solved, going from > 1h to 10 minutes seems like a good improvement.
